### PR TITLE
zot: run GC after manifest removal

### PR DESCRIPTION
**Pull request by tych0**
Clients today expect the repo to clean up if there are unused blobs, not to
manually delete things they think are unused. Let's do that, and use
umoci's code to do it since it's tested and works.

Signed-off-by: Tycho Andersen <tycho@tycho.ws>

### CODE REVIEW

**Comment by hallyn** [comment]
Now is that run locked against other actions on the same repo?  I.e. if someone else starts adding an image a moment after you've started the umoci.Clean for your delete, will you risk removing blobs which the add is expecting to be there (bc it had done a HEAD)?

**Comment by hallyn** [comment]
(Note that the distribution spec specifically says you should push the blobs first, then push the manifest referencing them, so the blobs could be GC'd in the meantime)

File: pkg/storage/storage_test.go [hunk 1]
```diff
@@ -135,8 +135,6 @@ func TestAPIs(t *testing.T) {
 					So(err, ShouldNotBeNil)
 				})
 
-				err = il.DeleteBlob("test", d.String())
-				So(err, ShouldBeNil)
```
Comment (rchincha) [inline pkg/storage/storage_test.go]: Since the GC is happening inline, this should actually be a non-nil error?

File: pkg/storage/storage.go [hunk 2]
```diff
@@ -502,6 +504,16 @@ func (is *ImageStore) DeleteImageManifest(repo string, reference string) error {
 		return err
 	}
 
+	oci, err := umoci.OpenLayout(dir)
+	if err != nil {
+		return err
+	}
+	defer oci.Close()
+
+	if err := oci.GC(context.Background()); err != nil {
```
Comment (rchincha) [inline pkg/storage/storage.go]: With dedupe, delete should be very fast since we are removing links for most part.

**Comment by tych0** [comment]
I think you're right that there are some races here. The only thing I can think is to lock while GC is running, and stop any parallel blob Put()s?

**Comment by hallyn** [comment]
So longer term, maybe the way to go about this is (a) keep all the blobs in one shared directory, hardlinking them into repo blob dirs as needed.  (b) when we gc, take all shared blobs which have nlink=1 and move them into purgatory/ ,  (c) when a blob in purgatory is checked for (HEAD) , move it back to shared, (d) when a manifest is uploaded, any blobs it refers to that are in purgatory are moved back, and then (d) after some time (an hour?) rm -rf jigoku; mv purgatory/ jigoku/ .  That way we won't keep any blobs around too long but avoid ever losing something that's in the process of being uploaded.

Shorter term what you (Tycho) have might work, I need to re-read.

**Comment by hallyn** [comment]
No, I don't think what you have works.  The problem is that we'd have to lock across the *client's* 'upload' routine, which from our point of view i just a bunch of HEAD gets followed by the manifest PUT.  Just locking those individually does not prevent the case of:

```
HEAD b1 -> ok
HEAD b2 -> absent
POST b2
HEAD b3 -> ok
[gc runs, b1 gets deleted]
POST manifest
```

**Comment by tych0** [comment]
Posting a manifest with non-present blobs will be rejected, won't it? So I
don't think this scenario is possible.

**Comment by hallyn** [comment]
> Posting a manifest with non-present blobs will be rejected, won't it? So I don't think this scenario is possible.

All right, yes, I guess that pkg/storage/storage.go:PutImageManifest() will return an ErrBlobNotFound in that case.  That's #notnice, and I don't think any client will not then assume that the repo must be busted, so the skopeo copy will fail - but at least the repo will be ok, so I guess it's sufficient for now.

**Comment by hallyn** [comment]
```
pkg/storage/storage.go:442:3: not enough arguments to return

	have (error)

	want (string, error)

pkg/storage/storage.go:447:3: not enough arguments to return

	have (error)

	want (string, error)
```

File: pkg/api/routes.go [hunk 3]
```diff
@@ -40,16 +41,25 @@ const (
 )
 
 type RouteHandler struct {
-	c *Controller
+	c        *Controller
+	blobLock sync.Mutex
```
Comment (rchincha) [inline pkg/api/routes.go]: yup, this is what I have in my PR as well, except just did a defer unlock() pattern.
A manifestLock is also needed coz multiple clients can potentially update/delete a manifest, and we need to decide a lock order in that case - manifest lock followed by blob lock.

**Comment by rchincha** [comment]
https://github.com/tych0/zot/pull/1

@tych0 to fix bazel build failures against this PR.

File: pkg/storage/storage_test.go (hunk 1 above)
Comment (rchincha) [reply pkg/storage/storage_test.go]: at least validates, GC did happen

File: pkg/api/routes.go (hunk 3 above)
Comment (tych0) [reply pkg/api/routes.go]: I'm not sure what defer unlock() does in terms of helping us out minimizing the critical section?

**Comment by tych0** [comment]
Hrm. The test failure looks legit; the GC is deleting the blob it's failing to open right before it. I must not understand how something in zot works: seems like the manifest should have a tag in the sub-oci-repo for that thing, right?

**Comment by rchincha** [comment]
This PR superseded by https://github.com/anuvu/zot/commit/bbed38cfa0c5481a2d91c2dee075ffe3ca151ad9.

**Pull request closed without merging**